virttest/qemu_vm.py

```diff
@@ -3561,34 +3561,38 @@ def activate_nic(self, nic_index_or_name):
         if nic['nic_model'] == 'virtio-net-pci':
             if int(nic['queues']) > 1:
                 device_add_cmd += ",mq=on"
-            if 'vectors' in nic:
-                device_add_cmd += ",vectors=%s" % nic.vectors
+                if 'vectors' in nic:
+                    device_add_cmd += ",vectors=%s" % nic.vectors
         device_add_cmd += nic.get('nic_extra_params', '')
         if 'romfile' in nic:
             device_add_cmd += ",romfile=%s" % nic.romfile
         bus = self.devices.get_buses({'aobject': 'pci.0'})[0]
-        if isinstance(bus, qdevices.QPCIEBus):
-            root_port_id = bus.get_free_root_port()
-            if root_port_id:
-                device_add_cmd += ",bus=%s" % root_port_id
-                root_port = self.devices.get_buses({'aobject': root_port_id})[0]
-                root_port.insert(qdevices.QBaseDevice(nic.nic_model,
-                                                      aobject=nic.nic_name))
-            else:
-                raise virt_vm.VMAddNicError("No free root port for device %s"
-                                            " to plug." % nic.nic_name)
-        error_context.context("Activating nic on VM %s with monitor command %s" % (
-            self.name, device_add_cmd))
+        nic_dev = qdevices.QDevice(params={'id': nic.nic_name,
+                                           'driver': nic.nic_model},
+                                   parent_bus=({'aobject': 'pci.0'},))
+        nic_dev.set_aid(nic_dev.get_qid())
+        bus.prepare_hotplug(nic_dev)
+        qdev_out = self.devices.insert(nic_dev)
+        if not isinstance(qdev_out, list) or len(qdev_out) != 1:
+            raise virt_vm.VMAddNicError("%s for device %s to plug."
+                                        % (qdev_out, nic.nic_name))
+        else:
+            device_add_cmd += ",bus=%s,addr=%s" % (nic_dev.get_param('bus'),
+                                                   nic_dev.get_param('addr'))
+
+        error_context.context("Activating nic on VM %s with monitor command %s"
+                              % (self.name, device_add_cmd))
 
         if self.monitor.protocol == 'qmp':
             self.monitor.send_args_cmd(device_add_cmd)
         else:
             self.monitor.send_args_cmd(device_add_cmd, convert=False)
 
         error_context.context("Verifying nic %s shows in qtree" % nic.nic_name)
-        qtree = self.monitor.info("qtree")
-        if nic.nic_name not in qtree:
+        qtree = self.monitor.info("qtree", debug=False)
```

Comment:
Directly checking if the ID exists does not seem to be very rigorous.  Suggest to use regex to match  special  line.

Reply:
Updated to a more rigorous check, thanks for suggestion.